Add form field ruleset

ericnewcomer opened the pull request:
* Adds new ruleset type for flow_field that requires a field index and delimiter
* Add config json to ruleset (webhook_url, etc.. should move here eventually)

nicpottier commented on temba/flows/models.py:
```diff
@@ -2493,6 +2513,24 @@ def find_matching_rule(self, step, run, msg):
+                elif delim == 'plus':
+                    delim = '\+'
+
+                fields = regex.split(delim, text)
```
unicode? I think there's a version flag you want to use as well. (check the other splits)

nicpottier commented on temba/flows/tests.py:
```diff
@@ -2294,6 +2296,46 @@ def clear_activity(self, flow):
+        assert_response("101+M+Seattle", "Sorry, 101 doesn't look like a valid age, please try again.")
+        assert_response("36+elephant+Seattle", "Sorry, elephant doesn't look like a valid gender. Try again.")
+        assert_response("36+M+Saturn", "I don't know the location Saturn. Please try again.")
+        assert_response("36+M+Seattle", "Thanks for your submission. We have that as:\n\n36 / M / Seattle")
```
Some unicode test cases would be good.

nicpottier commented on temba/flows/models.py:
```diff
@@ -2493,6 +2513,24 @@ def find_matching_rule(self, step, run, msg):
+                if index < len(fields):
+                    text = fields[index]
+                else:
+                    text = ''
```
Wonder if all the above should be a new expression.. something like ```FIELD(flow.field, "+", 3)``` .. that might allow more flexibility in the future.

ericnewcomer replied:
Also space isn't really space, it's greedy whitespace. But the same could be banned in here as well. I might be game as long as we keep this out of the flow definitions.

ericnewcomer replied:
Thinking a bit more on this. Should this field be direct string delimiters or slow regexes? Seems we'd ultimately want the flexibility of regexes here.  If that's the case it'll make the simple cases pretty weird. Especially since plus needs to be escaped and spaces probably should be greedy. 

If we don't make them regexes, then the plus case is natural like you have above but then spaces aren't greedy whitespace which is undesirable for the most common case. 

Encoding this as an operand expression seems fine, but I'd still argue for having it called out in the config in the definition so the UI isn't trying to deconstruct this stuff. I'd much rather the operand just hold the field and the ruleset logic constructs the expression at execution time from the config. Makes me wonder if it's really worth it, because it won't look much different in the end.

ericnewcomer replied:
@rowanseymour I'm curious of your take on this.

rowanseymour replied:
I think I like the idea of users entering a delimiter character and us using an expression function under the hood. 

What's a usecase that would need regexes? Most users are going to want to parse something like "A B   C" or "A + B  + C" right? So maybe just trim the parts after splitting and remove empties if splitting by space.

We already have a function that would almost do the job: `=WORD(text, number, by_spaces=False)`

nicpottier replied:
One thing to look at night be the stuff we had in nsms (or nsms-plus cant
remember which) for this kind of parsing. That was tweaked quite a bit
based on real life use cases by Hope and others so probably has some (now
forgotten) insights.

I dont think there's any need to base our decision on performance, that
isn't a real concern for this.
On Aug 5, 2015 10:41 PM, "Eric Newcomer" <notifications@github.com> wrote:

> In temba/flows/models.py
> <https://github.com/nyaruka/rapidpro/pull/282#discussion_r36382693>:
>
> > +
> > +                config = self.config_json()
> > +
> > +                # determine the proper regex to split on
> > +                delim = config.get('field_delimiter', 'space')
> > +                if delim == 'space':
> > +                    delim = '\\s+'
> > +                elif delim == 'plus':
> > +                    delim = '\+'
> > +
> > +                fields = regex.split(delim, text)
> > +                index = config.get('field_index', 0)
> > +                if index < len(fields):
> > +                    text = fields[index]
> > +                else:
> > +                    text = ''
>
> Thinking a bit more on this. Should this field be direct string delimiters
> or slow regexes? Seems we'd ultimately want the flexibility of regexes
> here. If that's the case it'll make the simple cases pretty weird.
> Especially since plus needs to be escaped and spaces probably should be
> greedy.
>
> If we don't make them regexes, then the plus case is natural like you have
> above but then spaces aren't greedy whitespace which is undesirable for the
> most common case.
>
> Encoding this as an operand expression seems fine, but I'd still argue for
> having it called out in the config in the definition so the UI isn't trying
> to deconstruct this stuff. I'd much rather the operand just hold the field
> and the ruleset logic constructs the expression at execution time from the
> config. Makes me wonder if it's really worth it, because it won't look much
> different in the end.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nyaruka/rapidpro/pull/282/files#r36382693>.
>

ericnewcomer replied:
Good point. After looking, the insight there is going with token based delimiters for splitting and stripping delimiters and whitespace afterwards.
https://github.com/nyaruka/nsms/blob/master/nsms/parser/parser.py#L49

The rest of it is things like next_phone, next_int, etc. which doesn't really apply in this case.

ericnewcomer commented:
Okay, I added a new function f_field and changed it so our form_field ruleset type constructs one of those internally. @rowanseymour care to have another look and give thoughts?

rowanseymour commented on temba/utils/parser_functions.py:
```diff
@@ -460,6 +460,27 @@ def f_word_slice(text, start, stop=0, by_spaces=False):
+    splits = text.split(delimiter)
+
+    # remove our delimiters and whitespace
+    splits = [ field for field in splits if field != delimiter and len(field.strip()) > 0 ]
```
Spaces in [  ] is a pep violation

ericnewcomer replied:
Your face is a pep violation

On Thu, Aug 6, 2015, 11:06 PM Rowan Seymour <notifications@github.com>
wrote:

> In temba/utils/parser_functions.py
> <https://github.com/nyaruka/rapidpro/pull/282#discussion_r36493107>:
>
> > @@ -460,6 +460,27 @@ def f_word_slice(text, start, stop=0, by_spaces=False):
> >      # re-combine selected words with a single space
> >      return ' '.join(selection)
> >
> > +def f_field(text, delimiter, index):
> > +    """
> > +    A field
> > +    :param text: the text to split
> > +    :param delimiter: the regex to split by
> > +    :param index: which index in the result to return
> > +    """
> > +
> > +    splits = text.split(delimiter)
> > +
> > +    # remove our delimiters and whitespace
> > +    splits = [ field for field in splits if field != delimiter and len(field.strip()) > 0 ]
>
> Spaces in [ ] is a pep violation
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nyaruka/rapidpro/pull/282/files#r36493107>.
>

rowanseymour commented on temba/utils/parser_functions.py:
```diff
@@ -460,6 +460,27 @@ def f_word_slice(text, start, stop=0, by_spaces=False):
 
+def f_field(text, delimiter, index):
+    """
+    A field
```
The docstring is supposed to be the help text that we display in the autocomplete... though I appreciate we'll need a different approach if we want localization

ericnewcomer replied:
Oops, forgot to fill that out properly, will fix

rowanseymour commented on temba/utils/parser_functions.py:
```diff
@@ -460,6 +460,27 @@ def f_word_slice(text, start, stop=0, by_spaces=False):
     # re-combine selected words with a single space
     return ' '.join(selection)
 
+def f_field(text, delimiter, index):
```
I'd change the order of the args to match `f_word(text, number, by_spaces=False)`. Perhaps also a case for having a default delimiter value.. e.g. f_field(text, index, delimiter=' ')

ericnewcomer replied:
Man that feels way less natural to me as a user, but consistency I guess. Suppose default as space makes sense, but that case is identical to word.  Feels a bit redundant. Hmmm.

rowanseymour replied:
"way less natural" is strong :)... if spaces is the common case then it just means you can write @FIELD(@step.value, 2) which is nice

The only real difference now is that WORD uses a "class of chars" as a delimiter and FIELD is using a specific char. I can't imagine anyone is actually using WORD.. could maybe ditch it if so.

ericnewcomer replied:
My words don't know their own strength.

rowanseymour replied:
Come think of it - `@WORD(step.value, N)` and `@FIRST_WORD(step.value)` primarily exist for compatibility with the `@step.value|first_word` and `@step.value|remove_first_word` filters, so we need them to behave as the currently do.

rowanseymour commented on temba/utils/parser_functions.py:
```diff
@@ -460,6 +460,27 @@ def f_word_slice(text, start, stop=0, by_spaces=False):
+    # remove our delimiters and whitespace
+    splits = [ field for field in splits if field != delimiter and len(field.strip()) > 0 ]
+
+    index = int(index)
```
For consistency this should be `val_to_integer`

rowanseymour replied:
If index < 1 raise a ValueError which will get bundled into a EvaluationError and will one day be reported back to the simulator

nicpottier commented:
Looks good to me, I like the new expression approach.
